File: frontend/src/lib/api/permissions.ts
import api from './client';

export interface PermissionEntry {
  module: string;
  action: string;
  is_allowed: boolean;
}

export interface PositionWithPerms {
  id: string;
  code: string;
  name: string;
  is_template: boolean;
  school_id: string | null;
  permissions: PermissionEntry[];
}

export const PERMISSION_MATRIX: Record<string, string[]> = {
  school:      ['view', 'edit', 'manage_users'],
  staff:       ['view', 'create', 'edit', 'delete'],
  students:    ['view', 'create', 'edit', 'delete'],
  academic:    ['view', 'create', 'edit', 'delete'],
  attendance:  ['view', 'record', 'approve'],
  assessments: ['view', 'enter_scores', 'approve_scores'],
  fees:        ['view', 'collect', 'manage'],
  housing:     ['view', 'assign', 'manage'],
  reports:     ['view', 'generate'],
};

export const MODULE_LABELS: Record<string, string> = {
  school:      'School',
  staff:       'Staff',
  students:    'Students',
  academic:    'Academic',
  attendance:  'Attendance',
  assessments: 'Assessments',
  fees:        'Fees',
  housing:     'Housing',
  reports:     'Reports',
};

export const ACTION_LABELS: Record<string, string> = {
  view:           'View',
  edit:           'Edit',
  create:         'Create',
  delete:         'Delete',
  manage_users:   'Manage users',
  record:         'Record',
  approve:        'Approve',
  enter_scores:   'Enter scores',
  approve_scores: 'Approve scores',
  collect:        'Collect',
  manage:         'Manage',
  assign:         'Assign',
  generate:       'Generate',
};

export const listPositionsWithPerms = () =>
  api.get<PositionWithPerms[]>('/permissions/positions').then(r => r.data);

export const updatePositionPermissions = (positionId: string, permissions: PermissionEntry[]) =>
  api.put<PositionWithPerms>(`/permissions/positions/${positionId}`, { permissions }).then(r => r.data);

/** Build a Set of "module.action" strings that are granted for a position */
export function grantedSet(pos: PositionWithPerms): Set<string> {
  return new Set(
    pos.permissions.filter(p => p.is_allowed).map(p => `${p.module}.${p.action}`)
  );
}

File: backend/app/schemas/permissions.py
from __future__ import annotations
import uuid
from pydantic import BaseModel, field_validator

# The canonical permission matrix — every module and its valid actions.
PERMISSION_MATRIX: dict[str, list[str]] = {
    "school":      ["view", "edit", "manage_users"],
    "staff":       ["view", "create", "edit", "delete"],
    "students":    ["view", "create", "edit", "delete"],
    "academic":    ["view", "create", "edit", "delete"],
    "attendance":  ["view", "record", "approve"],
    "assessments": ["view", "enter_scores", "approve_scores", "record_behaviour"],
    "fees":        ["view", "collect", "manage"],
    "housing":     ["view", "assign", "manage"],
    "reports":     ["view", "generate"],
    "documents":   ["view", "manage"],
    "lesson_plans": ["view", "manage"],
}

VALID_PAIRS: set[tuple[str, str]] = {
    (mod, act) for mod, acts in PERMISSION_MATRIX.items() for act in acts
}


class PermissionEntry(BaseModel):
    module: str
    action: str
    is_allowed: bool

    @field_validator("module")
    @classmethod
    def valid_module(cls, v: str) -> str:
        if v not in PERMISSION_MATRIX:
            raise ValueError(f"Unknown module '{v}'")
        return v

    @field_validator("action")
    @classmethod
    def valid_action(cls, v: str) -> str:
        return v


class PositionWithPerms(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    is_template: bool
    school_id: uuid.UUID | None
    permissions: list[PermissionEntry]
    model_config = {"from_attributes": True}


class PositionPermissionsUpdate(BaseModel):
    permissions: list[PermissionEntry]
